File: universal-utils/src/main/scala/io/github/ppdzm/utils/universal/base/SSH2Utils.scala
package io.github.ppdzm.utils.universal.base

import java.io.{BufferedReader, InputStream, InputStreamReader}

import ch.ethz.ssh2.{ChannelCondition, Connection, StreamGobbler}
import io.github.ppdzm.utils.universal.feature.LoanPattern

/**
 * @author Created by Stuart Alex on 2021/4/19.
 */
object SSH2Utils {

    def executeShellOnRemoteServer(remoteHost: String, user: String, password: String, command: String): Unit = {
        LoanPattern.using(new Connection(remoteHost)) {
            connection =>
                connection.connect()
                val isAuthenticated = (user == null && password == null) || connection.authenticateWithPassword(user, password)
                if (isAuthenticated) {
                    val session = connection.openSession
                    session.execCommand(command)
                    System.out.println("Following message is standard out:")
                    handleInputStream(session.getStdout)
                    System.out.println("Following message is standard err:")
                    handleInputStream(session.getStderr)
                    session.waitForCondition(ChannelCondition.EXIT_STATUS, 24 * 60 * 60)
                    session.close()
                }
        }
    }

    private def handleInputStream(inputStream: InputStream): Unit = {
        val gobbler = new StreamGobbler(inputStream)
        val reader = new BufferedReader(new InputStreamReader(gobbler))
        while (true) {
            val line = reader.readLine
            if (line == null)
                return
            System.out.println(line)
        }
    }

}

File: universal-utils/src/main/java/io/github/ppdzm/utils/universal/base/SSH2Utils.java
package io.github.ppdzm.utils.universal.base;

import ch.ethz.ssh2.ChannelCondition;
import ch.ethz.ssh2.Connection;
import ch.ethz.ssh2.Session;
import ch.ethz.ssh2.StreamGobbler;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;

/**
 * @author Created by Stuart Alex on 2021/6/11.
 */
@SuppressWarnings("AlibabaClassNamingShouldBeCamel")
public class SSH2Utils {

    public static void executeShellOnRemoteServer(String remoteHost, String user, String password, String command) throws IOException {
        Connection connection = new Connection(remoteHost);
        connection.connect();
        boolean isAuthenticated = (user == null && password == null) || connection.authenticateWithPassword(user, password);
        if (isAuthenticated) {
            Session session = connection.openSession();
            session.execCommand(command);
            System.out.println("Following message is standard out:");
            handleInputStream(session.getStdout());
            System.out.println("Following message is standard err:");
            handleInputStream(session.getStderr());
            session.waitForCondition(ChannelCondition.EXIT_STATUS, 24 * 60 * 60);
            session.close();
        }
        connection.close();
    }

    private static void handleInputStream(InputStream inputStream) throws IOException {
        StreamGobbler gobbler = new StreamGobbler(inputStream);
        BufferedReader reader = new BufferedReader(new InputStreamReader(gobbler));
        while (true) {
            String line = reader.readLine();
            if (line == null) {
                return;
            }
            System.out.println(line);
        }
    }
}
